shotover-proxy/src/transforms/redis/cache.rs

```diff
@@ -74,386 +296,493 @@ impl SimpleRedisCache {
+    }
+
+    /// clear the cache for the single row specified by the redis_key
+    fn clear_row_cache(
+        &mut self,
+        statement: &CassandraStatement,
+        table_cache_schema: &TableCacheSchema,
+    ) -> Option<Message> {
+        // TODO is it possible to return the future and process in parallel?
+
+        if let Ok((redis_key, _hash_key)) = build_redis_key_from_cql3(statement, table_cache_schema)
+        {
+            let commands_buffer: Vec<RedisFrame> = vec![
+                RedisFrame::BulkString("DEL".into()),
+                RedisFrame::BulkString(redis_key),
+            ];
+            Some(Message::from_frame(Frame::Redis(RedisFrame::Array(
+                commands_buffer,
+            ))))
+        } else {
+            None
         }
-        Ok(messages_cass_request)
     }
-}
 
-// TODO: We don't need to do it this way and allocate another struct
-struct ValueHelper(Value);
-
-impl ValueHelper {
-    fn as_bytes(&self) -> &[u8] {
-        match &self.0 {
-            Value::Number(v, false) => v.as_bytes(),
-            Value::SingleQuotedString(v) => v.as_bytes(),
-            Value::NationalStringLiteral(v) => v.as_bytes(),
-            Value::HexStringLiteral(v) => v.as_bytes(),
-            Value::Boolean(v) => {
-                if *v {
-                    &TRUE
+    /// calls the next transform and process the result for caching.
+    async fn execute_upstream_and_process_result<'a>(
+        &mut self,
+        message_wrapper: Wrapper<'a>,
+    ) -> ChainResponse {
+        let mut orig_messages = message_wrapper.messages.clone();
+        let orig_cql: Option<&mut CQL> = orig_messages
+            .iter_mut()
+            .filter_map(|message| {
+                if let Some(Frame::Cassandra(CassandraFrame {
+                    operation: CassandraOperation::Query { query, .. },
+                    ..
+                })) = message.frame()
+                {
+                    Some(query)
                 } else {
-                    &FALSE
+                    None
+                }
+            })
+            .next();
+        let result_messages = &mut message_wrapper.call_next_transform().await?;
+        if let Some(orig_cql) = orig_cql {
+            let mut cache_messages: Vec<Message> = vec![];
+            for (response, statement) in result_messages.iter_mut().zip(orig_cql.statements.iter())
+            {
+                if let Some(Frame::Cassandra(CassandraFrame {
+                    operation: CassandraOperation::Result(result),
+                    ..
+                })) = response.frame()
+                {
+                    match is_cacheable(statement) {
+                        CacheableState::Update { table_name }
+                        | CacheableState::Delete { table_name } => {
+                            if let Some(table_cache_schema) = self.caching_schema.get(&table_name) {
+                                let table_schema = table_cache_schema.clone();
```

Comment:
We can avoid this clone by rewriting as:
![image](https://user-images.githubusercontent.com/5120858/168511125-7279736b-001e-4f1b-a258-adbbbeeee1de.png)